File: compilers/ruby/test/test_directive_guard.rb
# frozen_string_literal: true

require_relative 'test_helper'

# The @sighash (#123) and @embedAlways (#109) comment directives are
# implemented only in the TypeScript compiler. The Ruby compiler must fail
# closed rather than silently drop them.
class TestDirectiveGuard < Minitest::Test
  def parse(source, file_name = 'Counter.runar.ts')
    RunarCompiler.send(:_parse_source, source, file_name)
  end

  def test_sighash_directive_is_rejected
    source = <<~TS
      class Counter extends SmartContract {
        readonly x: bigint;
        constructor(x: bigint) { super(x); this.x = x; }
        /** @sighash SINGLE|FORKID */
        public unlock() {}
      }
    TS
    result = parse(source)
    joined = result.error_strings.join("\n")
    assert result.errors.any?, 'expected fail-closed error for @sighash'
    assert_includes joined, '@sighash'
    assert_includes joined, '#123'
  end

  def test_embed_always_directive_is_rejected
    source = <<~TS
      class Counter extends SmartContract {
        /** @embedAlways */
        readonly x: bigint;
        constructor(x: bigint) { super(x); this.x = x; }
        public unlock() {}
      }
    TS
    result = parse(source)
    joined = result.error_strings.join("\n")
    assert result.errors.any?, 'expected fail-closed error for @embedAlways'
    assert_includes joined, '@embedAlways'
    assert_includes joined, '#109'
  end

  def test_non_directive_identifier_does_not_trip_guard
    # A field named `sighashType` must NOT trip the word-boundary guard.
    source = <<~TS
      class Counter extends SmartContract {
        readonly sighashType: bigint;
        constructor(sighashType: bigint) { super(sighashType); this.sighashType = sighashType; }
        public unlock() {}
      }
    TS
    result = parse(source)
    refute(result.error_strings.any? { |m| m.include?('not yet supported') },
           "directive guard tripped on a non-directive identifier: #{result.error_strings}")
  end
end

File: compilers/python/tests/test_directive_guard.py
"""Fail-closed directive guard (issues #123 @sighash / #109 @embedAlways).

The ``@sighash`` and ``@embedAlways`` comment directives are honoured only on
the TypeScript (.runar.ts) surface, which reads leading trivia. The non-TS
surface parsers ignore comments, so ``parse_source`` must FAIL CLOSED on those
formats rather than silently drop a security-critical directive. These tests
pin both halves of that policy.
"""

from __future__ import annotations

from runar_compiler.frontend.parser_dispatch import parse_source


def _errors(src: str, file_name: str) -> str:
    return "\n".join(parse_source(src, file_name).error_strings())


def test_ts_surface_honours_sighash_directive() -> None:
    src = (
        "class C extends StatefulSmartContract {\n"
        "  n: bigint;\n"
        "  constructor(n: bigint) { super(n); this.n = n; }\n"
        "  /** @sighash SINGLE|FORKID */\n"
        "  public bump(): void { this.n = this.n + 1n; }\n"
        "}\n"
    )
    r = parse_source(src, "C.runar.ts")
    assert not r.errors, r.error_strings()
    m = next(m for m in r.contract.methods if m.name == "bump")
    assert m.sighash_type == 0x43


def test_non_ts_surface_rejects_sighash_directive() -> None:
    src = (
        "contract Counter {\n"
        "  // @sighash SINGLE|FORKID\n"
        "  function unlock() public {}\n"
        "}\n"
    )
    joined = _errors(src, "Counter.runar.sol")
    assert "@sighash" in joined
    assert "#123" in joined


def test_non_ts_surface_rejects_embed_always_directive() -> None:
    src = (
        "module Counter {\n"
        "  // @embedAlways\n"
        "  x: u64;\n"
        "}\n"
    )
    joined = _errors(src, "Counter.runar.move")
    assert "@embedAlways" in joined
    assert "#109" in joined


def test_word_boundary_identifier_does_not_trip_guard() -> None:
    # A field named `sighashType` must NOT trip the `@sighash\b` guard.
    src = (
        "class C extends SmartContract {\n"
        "  readonly sighashType: bigint;\n"
        "  constructor(sighashType: bigint) { super(sighashType); this.sighashType = sighashType; }\n"
        "  public unlock(): void {}\n"
        "}\n"
    )
    joined = _errors(src, "C.runar.ts")
    assert "not supported" not in joined
